### DISCUSSION TITLE
다음 코드의 결과를 설명해주세요

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#100] [this]
- 다음 코드의 결과를 설명해주세요
```js
var globalObject = this;
var foo = () => this;
console.log(foo() === globalObject);

var obj = { func: foo };
console.log(obj.func() === globalObject);

console.log(foo.call(obj) === globalObject);

foo = foo.bind(obj);
console.log(foo() === globalObject);
```

**Comment by publdaze** [comment]
> * 다음 코드의 결과를 설명해주세요
> 
> ```js
> var globalObject = this;
> var foo = () => this;
> console.log(foo() === globalObject);
> 
> var obj = { func: foo };
> console.log(obj.func() === globalObject);
> 
> console.log(foo.call(obj) === globalObject);
> 
> foo = foo.bind(obj);
> console.log(foo() === globalObject);
> ```

- 첫번째 출력 - 화살표 함수 호출, 자신의 this 바인딩은 없고 상위 스코프 this 바인딩 사용. 상위스코프인 전역의 this는 globalObject와 동일하므로 `true`
- 두번째 출력 - 객체의 func 프로퍼티에 바인딩 되어 있는 foo 호출, 화살표 함수로 호출되므로 상위 스코프 this인 globalObject와 동일하므로 `true`
- 세번째, 네번째 출력 - 화살표 함수는 this 바인딩을 갖지 않기 때문에 call, apply, bind에 의해 this 변경되지 않음, 따라서 모두 `true`

**Comment by pipisebastian** [comment]
> * 다음 코드의 결과를 설명해주세요
> 
> ```js
> var globalObject = this;
> var foo = () => this;
> console.log(foo() === globalObject);
> 
> var obj = { func: foo };
> console.log(obj.func() === globalObject);
> 
> console.log(foo.call(obj) === globalObject);
> 
> foo = foo.bind(obj);
> console.log(foo() === globalObject);
> ```

- 모두 `true`를 반환합니다.
- 1번째 -> 일반함수로써 호출되나, 화살표 함수입니다.
  - 화살표 함수는, this를 바인딩하지 않습니다. 즉, 함수가 호출된 시점에서 this가 동적으로 정해지는 게 아닙니다. 그렇기 때문에, this를 참조하기 위해, 상위 스코프로 스코프 체이닝을 통해 찾게 되며, 이는 상위 스코프의 this를 따르게 됩니다. 이에 true를 반환합니다.
- 2번째 -> 객체의 메서드로써 호출되나, 마찬가지로 화살표함수이기에 true를 반환합니다
- 3,4번째 -> 명시적으로 this를 바인딩해주지만, 마찬가지로 화살표함수이기에 true를 반환합니다
